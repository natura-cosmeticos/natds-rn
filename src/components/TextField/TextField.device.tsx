import { storiesOf } from '@storybook/react-native'
import {
  Action,
  Default,
  Disabled,
  Feedback,
  Type,
  Readonly,
  Required,
  Sizes,
  States,
  TextArea,
  Interactive
} from './TextField.stories'

storiesOf('TextField', module)
  .add('Default', Default)
  .add('Sizes', Sizes)
  .add('States', States)
  .add('Feedback', Feedback)
  .add('Required', Required)
  .add('Disabled', Disabled)
  .add('Readonly', Readonly)
  .add('Action', Action)
  .add('Type', Type)
  .add('TextArea', TextArea)
  .add('Interactive', Interactive)
