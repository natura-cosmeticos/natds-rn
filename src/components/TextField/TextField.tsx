/* eslint-disable max-lines */
/* eslint-disable complexity */
/* eslint-disable no-unneeded-ternary */
import React, { useState } from 'react';
import { Text, TextInputProps } from 'react-native';
import { withTheme } from 'styled-components';
import { Theme } from '../../common/themeSelectors';

import {
  Wrapper,
  InputWrapper,
  Input,
  Label,
  HelperText,
} from './TextField.styles';

export type TextFieldStates = 'enabled' | 'focus' | 'active' | 'filled';

export interface TextFieldProps extends TextInputProps {
  /**
   * A disabled input is unusable
   */
  disabled?: boolean;
  /**
   * Style the input field according to the data  provided by the user to give a visual feedback
   * about the validation of the data
   */
  feedback?: 'error' | 'success';
  /**
   * Show a helper text below the input field
   */
  helperText?: string;
  /**
   * Descriptive text above the input field
   */
  label: string;
  /**
   * Define if the input has more than one line or not
   */
  multiline?: boolean;
  /**
   * Number of lines the input has
   */
  numberOfLines?: number;
  /**
   * onBlur event handler
   */
  onBlur?: (func) => void;
  /**
   * onChangeText event handler
   */
  onChangeText: (ev: string) => void;
  /**
   * onFocus event handler
   */
  onFocus?: (func) => void;
  /**
   * onSubmitEditing handler
   */
  onSubmitEditing?: () => void;
  /**
   * Input placeholder
   */
  placeholder: string;
  /**
   * When this prop is set, the value of the input can not be changed
   */
  readOnly?: boolean;
  /**
   * Show a visual indication if the input is required: Label*
   */
  required?: boolean;
  /**
   * Define the size of the input: tiny | small
   */
  size?: 'small' | 'regular';
  /**
   * TextField states: 'enabled' | 'focus' | 'active' | 'filled'
   */
  state?: TextFieldStates;
  /**
   * Id for testing
   */
  testID?: string;
  /**
   * app theme
   */
  theme: Theme;
  /**
   * Type of the input
   */
  type: 'text' | 'password';
  /**
   * Input value
   */
  value: string;
}

const TextFieldComponent = ({
  size = 'regular',
  testID = 'textField',
  required = false,
  disabled = false,
  readOnly = false,
  helperText = '',
  multiline = false,
  numberOfLines = 1,
  type,
  feedback,
  state,
  theme,
  label,
  placeholder,
  value,
  onBlur,
  onChangeText,
  onFocus,
  onSubmitEditing,
  ...props
}: TextFieldProps) => {
  const [currentState, setCurrentState] = useState<TextFieldStates>(() => {
    // If a state is provided, use it as the current state
    if (state) {
      return state;
    }

    // If value or readOnly is truthy, set the state as filled, else set as enabled
    return value || readOnly ? 'filled' : 'enabled';
  });

  const handleOnFocus = (func) => {
    setCurrentState('active');

    if (onFocus) {
      onFocus(func);
    }
  };

  const handleOnBlur = (func) => {
    setCurrentState(value ? 'filled' : 'enabled');

    if (onBlur) {
      onBlur(func);
    }
  };

  return (
    <Wrapper testID={testID}>
      {label !== '' && (
        <Label
          disabled={disabled}
          state={currentState}
          feedback={feedback}
          size={size}>
          <Text>{required ? `${label}*` : label}</Text>
        </Label>
      )}

      <InputWrapper
        disabled={disabled}
        state={currentState}
        size={size}
        testID={`${testID}-inputWrapper`}
        feedback={feedback}>
        <Input
          testID={`${testID}-input`}
          secureTextEntry={type === 'password'}
          onChangeText={onChangeText}
          placeholder={placeholder}
          size={size}
          value={value}
          multiline={multiline}
          numberOfLines={numberOfLines}
          onFocus={handleOnFocus}
          onBlur={handleOnBlur}
          onSubmitEditing={onSubmitEditing}
          editable={disabled || readOnly ? false : true}
          disabled={disabled}
          feedback={feedback}
          type={type}
          label={label}
          {...props}
        />
      </InputWrapper>

      <HelperText
        disabled={disabled}
        feedback={feedback}
        state={currentState}
        size={size}>
        {helperText !== '' && <Text>{helperText}</Text>}
      </HelperText>
    </Wrapper>
  );
};

export const TextField = withTheme(TextFieldComponent);
