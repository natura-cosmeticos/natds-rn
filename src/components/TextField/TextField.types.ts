/* eslint-disable max-lines */
/* eslint-disable max-len */
import { JSXElementConstructor } from 'react';
import {
  AccessibilityProps,
  Image,
  TextInputAndroidProps,
  TextInputIOSProps,
  TextInputProps,
} from 'react-native';
import { Size } from '@naturacosmeticos/natds-themes/react-native';
import { IconButtonProps } from '../IconButton/IconButton.types';

type AndroidProps = Pick<TextInputAndroidProps,
  | 'disableFullscreenUI'
  | 'importantForAutofill'
  | 'numberOfLines'
  | 'returnKeyLabel'
  | 'showSoftInputOnFocus'
  | 'textBreakStrategy'
> & {
  autoCompleteType?: Extract<TextInputAndroidProps, 'autoComplete'>
}

type IOSProps = Pick<TextInputIOSProps,
  | 'clearButtonMode'
  | 'clearTextOnFocus'
  | 'dataDetectorTypes'
  | 'enablesReturnKeyAutomatically'
  | 'keyboardAppearance'
  | 'passwordRules'
  | 'rejectResponderTermination'
  | 'scrollEnabled'
  | 'selectionState'
  | 'spellCheck'
  | 'textContentType'
>

type MultiPlatformProps = Pick<TextInputProps,
  | 'allowFontScaling'
  | 'autoCapitalize'
  | 'autoCorrect'
  | 'autoFocus'
  | 'blurOnSubmit'
  | 'contextMenuHidden'
  | 'defaultValue'
  | 'inputAccessoryViewID'
  | 'keyboardType'
  | 'maxFontSizeMultiplier'
  | 'maxLength'
  | 'multiline'
  | 'onBlur'
  | 'onChange'
  | 'onChangeText'
  | 'onContentSizeChange'
  | 'onEndEditing'
  | 'onFocus'
  | 'onKeyPress'
  | 'onScroll'
  | 'onSelectionChange'
  | 'onSubmitEditing'
  | 'onTextInput'
  | 'placeholder'
  | 'returnKeyType'
  | 'secureTextEntry'
  | 'selectTextOnFocus'
  | 'selection'
  | 'testID'
  | 'value'
>

/**
 * @deprecated Deprecated sizes: `small` and `regular, use 'medium' and 'mediumX' instead.
 */
type DeprecatedInputSizes = 'small' | 'regular';
export type InputSizes = keyof Pick<Size, 'medium' | 'mediumX'> | DeprecatedInputSizes
/**
 * @deprecated These are no longer necessary. This component now handles its own state.
 */
type TextFieldStates = 'enabled' | 'focus' | 'active' | 'filled';

export type TextFieldTypes = 'text' | 'password' | 'number';

interface BaseProps extends
  AccessibilityProps,
  MultiPlatformProps,
  AndroidProps,
  IOSProps {
    /**
     * Disabled TextField's are used when the user needs to perform
     * another action before being able to interact with the field.
     * When `true` the TextField is not editable and presents contents,
     * label, border and helper text in a less contrasting color.
     * @default false
     */
    disabled?: boolean;
    /**
     * When provided this shows above the field. It's used to identify
     * what the field is about.
     */
    label?: string;
    /**
     * Read only TextField's are used to show information the user can
     * not edit. When `true` the TextField is not editable and presents
     * a different background color.
     * @default false
     */
    readonly?: boolean;
    /**
     * When `true` the TextField's Label presents an asterisk after it.
     * Required TextField's are used when the user needs to fill it before moving
     * on to a next step.
     * @default false
     */
    required?: boolean;
    /**
     * This defines the height of the field, according to corresponding theme
     * Size token.
     *
     *  - Available sizes: `medium`, `mediumX`.
     *
     * ### Deprecated sizes: `small` and `regular`.
     * @default 'mediumX'
     */
    size?: InputSizes;
    /**
     * <span style="color:red">*Deprecated, this prop should not be used as providing states is no longer necessary, this component now handles its own state.*</span>
     *
     * Attribute responsible for changing the component's state.
     *
     *  - Available states: `enabled`, `focus`, `active`, `filled`.
     */
    state?: TextFieldStates;
    /**
     * This provides the most common behavior expected for this component without
     * the need to customize it.
     *
     * - Available types: `number`, `text`, `password`.
     * @default 'text'
     */
    type?: TextFieldTypes;
    /**
     * Determines which keyboard to open, e.g.numeric.
     * @default 'default'
     */
    keyboardType?: TextInputProps['keyboardType'];
    /**
     * If true, the text input obscures the text entered so that sensitive text like passwords stay secure.
     * @default false
     */
    secureTextEntry?: TextInputProps['secureTextEntry'];
  }

interface FeedbackProps {
  /**
   * When provided this shows the user visual feedback using bright
   * characteristic colors and icons. It's used to indicate validation
   * status.
   */
  feedback: 'error' | 'success';
  /**
   * When provided this shows below the field. It's used to provide
   * further information about how to fill in the field.
   */
  helperText: string;
}

interface WithoutFeedbackProps {
  feedback?: never;
  helperText?: string;
}

interface ActionImageProps {
  action: 'image';
  actionComponent: React.ReactComponentElement<JSXElementConstructor<Image>>;
}

interface ActionIconProps {
  action: 'icon';
  actionComponent: React.ReactComponentElement<JSXElementConstructor<IconButtonProps>>;
}

interface WithoutActionProps {
  action?: never;
  actionComponent?: never;
}

export type TextFieldProps = BaseProps
  & (
    | WithoutActionProps
    | ActionImageProps
    | ActionIconProps
  )
  & (
    | WithoutFeedbackProps
    | FeedbackProps
  )
