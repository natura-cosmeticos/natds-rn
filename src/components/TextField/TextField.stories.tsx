/* eslint-disable max-lines */
import React, { useState } from 'react';
import { View } from 'react-native';
import { select, text as textKnob, boolean as booleanKnob } from '@storybook/addon-knobs';
import { StoryContainer } from '../../common/HelperComponents/StoryContainer';
import { TextField } from './TextField';

export default {
  component: TextField,
  parameters: {
    componentSubtitle: 'TextField let users enter and edit text.',
  },
  title: 'Components|TextField',
};

const onChangeText = (ev: string) => {};
const onSubmitEditing = () => {};

export const Variants = () => {
  const [value, setValue] = useState('');

  return (
    <StoryContainer title="Variants">
      <TextField
        type="text"
        onChangeText={(ev: string) => setValue(ev)}
        onSubmitEditing={onSubmitEditing}
        label="Label"
        placeholder="Placeholder"
        value={value}
        helperText="Helper Text"
      />
    </StoryContainer>
  );
};

export const Size = () => {
  const [firstValue, setFirstValue] = useState('');
  const [secondValue, setSecondValue] = useState('');

  return (
    <StoryContainer title="Sizes">
      <View style={{ flex: 1, flexDirection: 'column' }}>
        <TextField
          type="text"
          onChangeText={(ev: string) => setFirstValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={firstValue}
          size="small"
          helperText="Helper Text"
        />

        <TextField
          type="text"
          onChangeText={(ev: string) => setSecondValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={secondValue}
          helperText="Helper Text"
        />
      </View>
    </StoryContainer>
  );
};

export const State = () => {
  const [firstValue, setFirstValue] = useState('');
  const [secondValue, setSecondValue] = useState('');
  const [thirdValue, setThirdValue] = useState('Hello World');

  return (
    <StoryContainer title="States">
      <View style={{ flex: 1, flexDirection: 'column' }}>
        <TextField
          type="text"
          onChangeText={(ev: string) => setFirstValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Enabled"
          placeholder="Placeholder"
          value={firstValue}
          state="enabled"
          helperText="Helper Text"
        />

        <TextField
          type="text"
          onChangeText={(ev: string) => setSecondValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Active"
          placeholder="Placeholder"
          value={secondValue}
          state="active"
          helperText="Helper Text"
        />

        <TextField
          type="text"
          onChangeText={(ev: string) => setThirdValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Filled"
          placeholder="Placeholder"
          value={thirdValue}
          state="filled"
          helperText="Helper Text"
        />
      </View>
    </StoryContainer>
  );
};

export const Feedback = () => {
  const [firstValue, setFirstValue] = useState('');
  const [secondValue, setSecondValue] = useState('');

  return (
    <StoryContainer title="Feedback">
      <View style={{ flex: 1, flexDirection: 'column' }}>
        <TextField
          type="text"
          onChangeText={(ev: string) => setFirstValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={firstValue}
          feedback="success"
          helperText="Helper Text"
        />

        <TextField
          type="text"
          onChangeText={(ev: string) => setSecondValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={secondValue}
          feedback="error"
          helperText="Helper Text"
        />
      </View>
    </StoryContainer>
  );
};

export const Required = () => {
  const [value, setValue] = useState('');

  return (
    <StoryContainer title="Required">
      <TextField
        type="text"
        onChangeText={(ev: string) => setValue(ev)}
        onSubmitEditing={onSubmitEditing}
        label="Label"
        placeholder="Placeholder"
        value={value}
        required
        helperText="Helper Text"
      />
    </StoryContainer>
  );
};

export const Disabled = () => (
  <StoryContainer title="Disabled">
    <TextField
      type="text"
      onChangeText={onChangeText}
      onSubmitEditing={onSubmitEditing}
      label="Label"
      placeholder="Placeholder"
      value=""
      disabled
      helperText="Helper Text"
    />
  </StoryContainer>
);

export const ReadOnly = () => (
  <StoryContainer title="Read Only">
    <TextField
      type="text"
      onChangeText={onChangeText}
      onSubmitEditing={onSubmitEditing}
      label="Label"
      placeholder="Placeholder"
      value="Hello World"
      readOnly
      helperText="Helper Text"
    />
  </StoryContainer>
);

export const HelperText = () => {
  const [firstValue, setFirstValue] = useState('');
  const [secondValue, setSecondValue] = useState('');

  return (
    <StoryContainer title="Helper Text">
      <View style={{ flex: 1, flexDirection: 'column' }}>
        <TextField
          type="text"
          onChangeText={(ev: string) => setFirstValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={firstValue}
          helperText="Helper Text"
        />

        <TextField
          type="text"
          onChangeText={(ev: string) => setSecondValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={secondValue}
        />
      </View>
    </StoryContainer>
  );
};

export const Type = () => {
  const [firstValue, setFirstValue] = useState('Text');
  const [secondValue, setSecondValue] = useState('Password');

  return (
    <StoryContainer title="Type">
      <View style={{ flex: 1, flexDirection: 'column' }}>
        <TextField
          type="text"
          onChangeText={(ev: string) => setFirstValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={firstValue}
          helperText="Helper Text"
        />

        <TextField
          type="password"
          onChangeText={(ev: string) => setSecondValue(ev)}
          onSubmitEditing={onSubmitEditing}
          label="Label"
          placeholder="Placeholder"
          value={secondValue}
          helperText="Helper Text"
        />
      </View>
    </StoryContainer>
  );
};

export const Interactive = () => {
  const [value, setValue] = useState('');

  return (
    <StoryContainer title="Interactive">
      <TextField
        type={select('Types', { password: 'password', text: 'text' }, 'text')}
        onChangeText={(ev: string) => setValue(ev)}
        onSubmitEditing={onSubmitEditing}
        label={textKnob('Label', 'Label')}
        placeholder={textKnob('Placeholder', 'Placeholder')}
        value={value}
        helperText={textKnob('HelperText', 'HelperText')}
        multiline={booleanKnob('Multiline', false)}
      />
    </StoryContainer>
  );
};

export const All = () => (
  <View>
    <Variants />
    <Size />
    <State />
    <Feedback />
    <Required />
    <Disabled />
    <ReadOnly />
    <HelperText />
    <Type />
  </View>
);
