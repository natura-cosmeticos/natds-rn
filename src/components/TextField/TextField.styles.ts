import { Theme } from '@naturacosmeticos/natds-themes/react-native';
import { CSSObject, CSSProperties } from 'styled-components';
import styled from 'styled-components/native';
import {
  buildColorWithOpacity,
  getBorderRadiusMedium,
  getColorHighEmphasis,
  getColorLowEmphasis,
  getColorMediumEmphasis,
  getColorSurface,
  getOpacityDisabledLow,
  getSizeLarge,
  getSizeMedium,
  getSizeMediumX,
  getSpacingSmall,
  getSpacingTiny,
} from '../../common/themeSelectors';
import { InputSizes } from './TextField.types';

interface InputStyleProps {
  disabled: boolean,
  hasActionIcon: boolean,
  readonly: boolean,
  size: InputSizes,
  theme: Theme,
}
const isSizeMediumX = (size: InputSizes) => size === 'mediumX';
const getFieldHeight = (size: InputSizes, theme: Theme) => (
  isSizeMediumX(size) ? getSizeMediumX(theme) : getSizeMedium(theme)
);
const getReadonlyColor = (theme: Theme) => buildColorWithOpacity(
  getColorLowEmphasis, getOpacityDisabledLow, theme,
);
const getTextColor = (disabled: boolean, theme: Theme) => (
  disabled ? getColorLowEmphasis(theme) : getColorHighEmphasis(theme)
);

export const getPlaceholderTextColor = (disabled: boolean, theme: Theme) => (
  disabled ? getColorLowEmphasis(theme) : getColorMediumEmphasis(theme)
);

export const Input = styled.TextInput<InputStyleProps & CSSProperties>(({
  disabled, hasActionIcon, readonly, size, theme,
}): CSSObject => ({
  backgroundColor: readonly ? getReadonlyColor(theme) : getColorSurface(theme),
  borderRadius: getBorderRadiusMedium(theme),
  color: getTextColor(disabled, theme),
  flexGrow: 1,
  maxWidth: '100%',
  minHeight: getFieldHeight(size, theme),
  overflow: 'hidden',
  paddingLeft: getSpacingSmall(theme),
  paddingRight: hasActionIcon ? 0 : getSpacingSmall(theme),
  width: 10,
}));

export const ActionImage = styled.View<Pick<InputStyleProps, 'size' | 'theme'>>(
  ({ size, theme }): CSSObject => ({
    borderBottomRightRadius: getBorderRadiusMedium(theme),
    borderTopRightRadius: getBorderRadiusMedium(theme),
    maxHeight: getFieldHeight(size, theme),
    overflow: 'hidden',
    width: getSizeLarge(theme),
  }),
);

export const ActionIcon = styled.View(({ theme }): CSSObject => ({
  paddingHorizontal: getSpacingTiny(theme),
}));
