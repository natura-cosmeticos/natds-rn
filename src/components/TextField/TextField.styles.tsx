/* eslint-disable max-lines */

import styled from 'styled-components/native';
import {
  Theme,
  getColorAlert,
  getColorHighEmphasis,
  getColorLowEmphasis,
  getColorMediumEmphasis,
  getColorPrimary,
  getColorSuccess,
  getSize,
} from '../../common/themeSelectors';

import { TextFieldStates, TextFieldProps as TextField } from './TextField';

interface TextFieldProps {
  feedback: string | undefined;
  theme: Theme;
  state: TextFieldStates;
  disabled: boolean;
  size?: 'small' | 'regular';
  numberOfLines?: number;
}

interface AttrsProps {
  theme: Theme;
  disabled: boolean;
}

const disabledStyles = (theme: Theme) => {
  const style = {
    border: '1px',
    borderColor: getColorLowEmphasis(theme),
    color: getColorLowEmphasis(theme),
  };

  return style;
};

const feedbackStyles = (theme: Theme, feedback: string) => {
  const style = {
    error: {
      border: '2px',
      borderColor: getColorAlert(theme),
      color: getColorAlert(theme),
    },
    success: {
      border: '2px',
      borderColor: getColorSuccess(theme),
      color: getColorSuccess(theme),
    },
  };

  return style[feedback];
};

const stateStyles = (theme: Theme, state: TextFieldStates) => {
  const style = {
    active: {
      border: '2px',
      borderColor: getColorPrimary(theme),
      color: getColorMediumEmphasis(theme),
    },
    enabled: {
      border: '1px',
      borderColor: getColorLowEmphasis(theme),
      color: getColorMediumEmphasis(theme),
    },
    filled: {
      border: '1px',
      borderColor: getColorHighEmphasis(theme),
      color: getColorMediumEmphasis(theme),
    },
    focus: {
      border: '2px',
      borderColor: getColorPrimary(theme),
      color: getColorMediumEmphasis(theme),
    },
  };

  return style[state];
};

const getInputStyles = (
  state: TextFieldStates,
  feedback: string | undefined,
  disabled: boolean,
  theme: Theme,
) => {
  if (disabled) {
    const style = disabledStyles(theme);

    return style;
  }

  if (feedback) {
    const style = feedbackStyles(theme, feedback);

    return style;
  }

  const style = stateStyles(theme, state);

  return style;
};

const getHeight = (size, theme, numberOfLines) => {
  const sizeName = size === 'regular' ? 'mediumX' : 'medium';
  const sizeInPixels = getSize(theme, sizeName);

  if (numberOfLines) {
    return numberOfLines * sizeInPixels;
  }

  return sizeInPixels;
};

export const Wrapper = styled.View`
  width: 100%;
  margin-bottom: 2px;
`;

export const Label = styled.Text<TextFieldProps>`
  font-size: 14px;
  margin-bottom: 5px;

  color: ${({
    state,
    feedback,
    disabled,
    theme,
  }) => getInputStyles(state, feedback, disabled, theme).color};
`;

export const InputWrapper = styled.View<TextFieldProps>`
  border: ${({
    state,
    feedback,
    disabled,
    theme,
  }) => getInputStyles(state, feedback, disabled, theme).border};

  border-color: ${({
    state,
    feedback,
    disabled,
    theme,
  }) => getInputStyles(state, feedback, disabled, theme).borderColor};


  padding: 8px 16px;
  height: ${({ theme, size = 'regular', numberOfLines }) => getHeight(size, theme, numberOfLines)}px;
  border-radius: 4px;
`;

export const Input = styled.TextInput.attrs<AttrsProps>(
  ({ theme, disabled }) => ({
    placeholderTextColor: disabled
      ? getColorLowEmphasis(theme)
      : getColorMediumEmphasis(theme),
  }),
)<TextField>`
  font-size: 16px;
  flex: 1;
  margin-right: ${({ type }) => (type === 'password' ? '40px' : '0')};
  text-align-vertical: ${({ multiline }) => (multiline ? 'top' : 'center')};
  height: ${({ theme, size = 'regular', numberOfLines }) => getHeight(size, theme, numberOfLines)}px;
  color: ${({ theme }) => getColorHighEmphasis(theme)};
`;

export const IconPress = styled.TouchableOpacity`
  align-self: center;
  margin-right: 16px;
`;

export const HelperText = styled.Text<TextFieldProps>`
  margin-top: 5px;
  font-size: 12px;
  min-height: 20px;
  color: ${({
    state,
    feedback,
    disabled,
    theme,
  }) => getInputStyles(state, feedback, disabled, theme).color};
`;
