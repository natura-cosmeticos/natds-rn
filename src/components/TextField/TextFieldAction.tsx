import React from 'react';
import { View } from 'react-native';
import { IconButton } from '../IconButton';
import { ActionIcon, ActionImage } from './TextField.styles';
import { AccessibilityButtonProps } from './TextField.types';

type TextFieldActionStyleProps = Pick<AccessibilityButtonProps,
  'accessibilityButtonHint' |
  'accessibilityButtonLabel' |
  'accessibilityButtonRole'>
& {
  secureState: boolean,
  onPress: () => void
}

export const TextFieldPasswordAction = ({
  accessibilityButtonHint,
  accessibilityButtonLabel,
  secureState,
  onPress,
}: TextFieldActionStyleProps) => (
  <ActionIcon testID="password-action-icon" >
    <IconButton
      accessibilityHint={accessibilityButtonHint}
      accessibilityLabel={accessibilityButtonLabel}
      onPress={onPress}
      icon={secureState
        ? 'outlined-action-visibilityoff'
        : 'outlined-action-visibility'} />
  </ActionIcon>
);

export const TextFieldAction = ({
  action,
  actionComponent,
  size,
  accessibilityButtonHint,
  accessibilityButtonLabel,
  accessibilityButtonRole,
}) => (
  <View>
    { action === 'icon'
      && <ActionIcon
          testID="action-icon"
          accessibilityHint={accessibilityButtonHint}
          accessibilityLabel={accessibilityButtonLabel}
          accessibilityRole={accessibilityButtonRole}
        >
          { actionComponent }
        </ActionIcon>
    }
    { action === 'image'
      && <ActionImage
          accessibilityHint={accessibilityButtonHint}
          accessibilityLabel={accessibilityButtonLabel}
          accessibilityRole={accessibilityButtonRole}
          testID="action-image"
          size={size}
        >
          { actionComponent }
        </ActionImage>
    }
  </View>
);
