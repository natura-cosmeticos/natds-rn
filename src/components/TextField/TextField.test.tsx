import { fireEvent } from '@testing-library/react-native';
import React from 'react';
import { renderWithTheme } from '../../../test/testHelpers';
import { TextField } from './TextField';
// @ts-ignore
import placeholderImage from '../../assets/images/anonymous.jpg';

jest.mock('../../common/themeSelectors', () => (
  {
    buildColorWithOpacity: () => '#BBBBBB00',
    getBorderRadiusMedium: () => 8,
    getColorAlert: () => '#FF0000',
    getColorByName: () => '#0F0F0F',
    getColorHighEmphasis: () => '#333333',
    getColorHighlight: () => '#333333',
    getColorLowEmphasis: () => '#BBBBBB',
    getColorMediumEmphasis: () => '#777777',
    getColorPrimary: () => '#FFAA33',
    getColorSuccess: () => '#00FF00',
    getColorSurface: () => '#000000',
    getOpacityDisabledLow: () => 0.5,
    getSize: () => 16,
    getSizeLarge: () => 42,
    getSizeMedium: () => 24,
    getSizeMediumX: () => 32,
    getSizeSemi: () => 14,
    getSpacingMicro: () => 4,
    getSpacingSmall: () => 8,
    getSpacingTiny: () => 2,
    getTheme: () => {},
    getTypographyStyles: () => ({
      caption: {}, subtitle2: {},
    }),
  }));

describe('TextField', () => {
  describe('render', () => {
    it('should render with default props', () => {
      const { getByTestId, toJSON } = renderWithTheme(<TextField />);

      const defaultStyles = {
        backgroundColor: '#000000',
        color: '#333333',
        minHeight: 32,
      };

      expect(getByTestId('ds-input')).toHaveStyle(defaultStyles);
      expect(toJSON()).toMatchSnapshot();
    });
    it('should render with given size', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" size="medium" />);

      expect(getByTestId('input')).toHaveStyle({ minHeight: 24 });
    });
    it('should render with correct props when disabled', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" disabled />);

      expect(getByTestId('input')).toHaveStyle({ color: '#BBBBBB' });
      expect(getByTestId('input')).toHaveProp('editable', false);
    });
    it('should render with correct props when readonly', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" readonly />);

      expect(getByTestId('input')).toHaveStyle({ backgroundColor: '#BBBBBB00' });
      expect(getByTestId('input')).toHaveProp('editable', false);
    });
  });
  describe('content', () => {
    it('should get given value', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" value="testing value" />);

      expect(getByTestId('input')).toHaveProp('value', 'testing value');
    });
    it('should get given placeholder', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" placeholder="testing placeholder" />);

      expect(getByTestId('input')).toHaveProp('placeholder', 'testing placeholder');
    });
  });
  describe('status events', () => {
    it('should call onFocus when focused', () => {
      const onFocusMock = jest.fn();
      const { getByTestId } = renderWithTheme(<TextField testID="input" onFocus={onFocusMock} />);

      fireEvent(getByTestId('input'), 'focus');

      expect(onFocusMock).toHaveBeenCalled();
    });
    it('should call onBlur when loses focus', () => {
      const onBlurMock = jest.fn();
      const { getByTestId } = renderWithTheme(<TextField testID="input" onBlur={onBlurMock} />);

      fireEvent(getByTestId('input'), 'blur');

      expect(onBlurMock).toHaveBeenCalled();
    });
    it('should call changeText function with value', () => {
      const mockChange = jest.fn();
      const { getByTestId } = renderWithTheme(<TextField testID="input" onChangeText={mockChange} />);

      const testText = 'testing change';

      fireEvent(getByTestId('input'), 'onChangeText', testText);
      expect(mockChange).toHaveBeenCalledWith(testText);
    });
  });
  describe('action', () => {
    it('should render an icon button when receives action icon', () => {
      const { getByTestId } = renderWithTheme(
        <TextField action="icon" actionComponent={<></>} />,
      );

      expect(getByTestId('action-icon')).toBeTruthy();
    });
    it('should render an image button when receives action image', () => {
      const { getByTestId } = renderWithTheme(
        <TextField action="image" actionComponent={<></>} />,
      );

      expect(getByTestId('action-image')).toBeTruthy();
    });
  });
  describe('type password', () => {
    it('should render with correct props when type is password', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" type="password" />);
      const iconName = getByTestId('password-action-icon').props.children.props.icon;

      expect(iconName).toBe('outlined-action-visibilityoff');
      expect(getByTestId('input')).toHaveProp('secureTextEntry', true);
    });
    it('should handle value visibility when type is password', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" type="password" />);

      fireEvent(getByTestId('password-action-icon'), 'press');
      const iconName = getByTestId('password-action-icon').props.children.props.icon;

      expect(iconName).toBe('outlined-action-visibility');
      expect(getByTestId('input')).toHaveProp('secureTextEntry', false);
    });
  });
  describe('type number', () => {
    it('should show keyboard type number when active', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" type="number" />);

      expect(getByTestId('input')).toHaveProp('keyboardType', 'numeric');
    });
    it('should only accept numbers', () => {
      const { getByTestId } = renderWithTheme(<TextField testID="input" type="number" />);

      fireEvent(getByTestId('input'), 'changeText', '1234abc567');

      expect(getByTestId('input')).toHaveProp('value', '1234567');
    });
  });
});
