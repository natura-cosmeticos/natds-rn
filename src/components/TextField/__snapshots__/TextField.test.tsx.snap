// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField render should render with default props 1`] = `
<View
  style={
    Array [
      Object {
        "width": "100%",
      },
    ]
  }
>
  <View
    borderWidth={1}
    style={
      Array [
        Object {
          "borderColor": "transparent",
          "borderRadius": 8,
          "borderWidth": 1,
        },
      ]
    }
    testID="box-container"
  >
    <View
      borderColor="#BBBBBB"
      borderWidth={1}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#000000",
            "borderColor": "#BBBBBB",
            "borderRadius": 8,
            "borderWidth": 1,
            "flexDirection": "row",
            "maxWidth": "100%",
          },
        ]
      }
      testID="box"
    >
      <TextInput
        allowFontScaling={true}
        disabled={false}
        editable={true}
        hasActionIcon={false}
        keyboardType="default"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#777777"
        readonly={false}
        rejectResponderTermination={true}
        secureTextEntry={false}
        size="mediumX"
        style={
          Array [
            Object {
              "backgroundColor": "#000000",
              "borderRadius": 8,
              "color": "#333333",
              "flexGrow": 1,
              "height": 32,
              "maxWidth": "100%",
              "overflow": "hidden",
              "paddingLeft": 8,
              "paddingRight": 8,
              "width": 10,
            },
          ]
        }
        testID="input"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
</View>
`;
